ChainSafe/gossamer
# fix(dot/sync, dot/rpc): implement HighestBlock

## Changes
- dot:
	- Create createRPCServiceParams to utilize Parameter Object Pattern
		- Update node.go to utilize createRPCServiceParams
- dot/rpc/modules:
	- Create SyncAPI interface and remove HighestBlock() from NetworkAPI
	- Include SyncAPI interface in SystemModule
	- Calling SyncAPI.HighestBlock() in SystemModule.SyncState() instead of NetworkAPI.HighestBlock()
- dot/rpc:
	- Include SyncAPI interface in HTTPServerConfig
- dot/sync:
	- include getHighestBlock() in ChainSync interface
	- implement HighestBlock() in Sync.Service
	- getHighestBlock() uses strategy to include peers numbers in a slice > sort this slice > get first value.

## Tests

- dot/rpc/modules/mock:
	- dot/rpc/modules/mock: 
		- Generate mock for SyncAPI and update NetworkAPI
	- dot/sync: 
		- Create TestHighestBlock
		- Test implementaion for *chainSync.getHighestBlock

* `cd dot/sync && go test -run TestChainSync_highestBlock`
* `cd dot/sync && go test -run TestHighestBlock`
* I recommend to run others test which was modified in this PR just to connect with SyncAPI interface.

@noot this connection is working, however, I would like to check it in my dev environment. How could I do that?

What I have tried:
* build with: `make build`
* Up my dev environment with: `./bin/gossamer --chain dev init && ./bin/gossamer --chain dev`
* Exec: `curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "system_syncState", "params":[]}' http://localhost:8545`

Do I need to change this ID? If I do, for what? 
How could I test it? Do u have any idea?

## Issues

- #1857

## Primary Reviewer

@noot
